src/magic_folder/test/test_local_snapshot.py

```diff
@@ -0,0 +1,336 @@
+
+        with ExpectedException(ValueError,
+                               "expected a file"):
+            self.snapshot_service.add_file(bar_dir)
+
+
+        # try adding a file outside the magic folder directory
+        tmpfile = FilePath(self.mktemp())
+        with tmpfile.open("wb") as f:
+            f.write(content)
+
+        with ExpectedException(ValueError,
+                               "The path being added .*"):
+            self.snapshot_service.add_file(tmpfile)
+
+        self.assertThat(
+            self.snapshot_service.stopService(),
+            succeeded(Always())
+        )
+
+    @given(content1=binary(min_size=1),
+           content2=binary(min_size=1),
+           filename=path_segments().map(lambda p: p.encode("utf-8")),
+    )
+    def test_add_a_file_twice(self, filename, content1, content2):
+        """
+        Adding a file twice should create two local snapshots with the first
+        one being the parent of the second.
+        """
+        foo = self.magic_path.child(filename)
+        with foo.open("wb") as f:
+            f.write(content1)
+
+        self.snapshot_service.startService()
+        self.snapshot_service.add_file(foo)
+
+        with foo.open("wb") as f:
+            f.write(content2)
+
+        # it should use the previous localsnapshot as its parent.
+        self.snapshot_service.add_file(foo)
+
+        self.assertThat(
+            self.snapshot_service.stopService(),
+            succeeded(Always())
+        )
+
+        self.assertThat(self._snapshot_creator.processed.get(foo),
+                        Equals(1)
+        )
+
+class LocalSnapshotCreatorTests(SyncTestCase):
+    """
+    Tests for LocalSnapshotCreator that is responsible for
+    creating the local snapshots and storing them in the database.
+    """
+    def setUp(self):
+        # setup db
+        # create author
+        # setup stashdir
+        # setup magicpath dir (the base directory for a particular magic folder)
+        # instantiate LocalSnapshotCreator
+        super(LocalSnapshotCreatorTests, self).setUp()
+        self.db = magicfolderdb.get_magicfolderdb(":memory:", create_version=(magicfolderdb.SCHEMA_v1, 1))
+        self.author = create_local_author("alice")
+
+        self.stash_dir = self.mktemp()
+        os.mkdir(self.stash_dir)
+
+        magic_path_dirname = self.mktemp()
+        os.mkdir(magic_path_dirname)
+        self.magic_path = FilePath(magic_path_dirname)
+
+        self.snapshot_creator = LocalSnapshotCreator(
+            db=self.db,
+            author=self.author,
+            stash_dir=FilePath(self.stash_dir),
+        )
+
+    def setup_example(self):
+        """
+        Hypothesis-invoked hook to create per-example state.
+        Reset the database before running each test.
+        """
+        # XXX: Not sure if this function should exist at all and even if
+        # so, it shouldn't exist in magicfolderdb module perhaps?
+        self.db._clear_snapshot_table()
```

Comment:
It's probably fine for the short while longer that magicfolderdb is going to exist.